metals/src/main/scala/scala/meta/internal/metals/MetalsEnrichments.scala

```diff
@@ -352,6 +380,40 @@ object MetalsEnrichments
         }
         AbsolutePath(out)
       }
+
+      if (path.toNIO.getFileSystem == workspace.toNIO.getFileSystem) {
+        path
+      } else {
+        path.jarPath match {
+          case Some(jar) =>
+            val jarDir =
+              workspace.resolve(Directories.dependencies).resolve(jar.filename)
+            val out = jarDir.resolveZipPath(path.toNIO)
+            val jarMetaFile = jarDir.resolve("jar.meta")
+
+            lazy val currentJarMeta = readJarMeta(jarMetaFile)
+            lazy val jarMeta = toJarMeta(jar)
+
+            val updateMeta = !jarDir.exists || !currentJarMeta.contains(jarMeta)
+            if (!out.exists || updateMeta) {
+              withJarDirLock(jarDir) {
+                if (updateMeta) {
+                  val prevFiles = FileIO
+                    .listAllFilesRecursively(jarDir)
+                    .filter(_.filename != ".lock")
+                  prevFiles.foreach(_.delete())
+                  Files.write(jarMetaFile.toNIO, jarMeta.getBytes)
+                }
+                copyFile(path, out)
+              }(retry)
```

Comment:
Probably not a real issues, but could this ever cause a loop? If there is an `IOException`, then if we wait a bit and try again it should be fine, but is there any situation where that exception could keep happening and fall into an infinite loop?

Reply:
Good point. I think there non-zero chance to fall into it due to some extra obstacles like unexpected workspace directory ownership changes  /not enough space on disk etc.
Added an exit clause to avoid infinity loop.